R/generalized_normal.R

```diff
@@ -0,0 +1,188 @@
+    stop("generalized_normal argument vectors could not be matched. May be due to wrong type,
+         or different lengths. Note: len=1 is always allowed, even if the other vectors are len!=1.")
+  }
+  if (!isLogic_len(log)) {
+    stop("the log argument of a density has to be a scalar boolean")
+  }
+
+
+  z <- (x - mu) / sigma
+  lpdf <- log(beta) - (log(2) + log(sigma) + lgamma(1/beta)) - abs(abs(x - mu) / sigma)^beta
+
+  #return either the log or the pdf itself, given the log-value
+  if (log) {
+    return(lpdf)
+  } else {
+    return(exp(lpdf))
+  }
+}
+
+#' Quantile function of the generalized_normal distribution
+#'
+#' @param p quantile value, 0 < p < 1
+#' @param mu Mean, unbound
+#' @param sigma Scale, sigma > 0
+#' @param beta shape, beta > 0
+#'
+#' @return Quantiles of the generalized_normal distribution
+#' @export
+#'
+#' @examples x <- seq(from = 0.01, to = 0.99, length.out = 100)
+#' plot(x, qgeneralized_normal(x, mu = 2, sigma = 2, beta = 1), type = "l")
+qgeneralized_normal <- function(p, mu, sigma, beta) {
+  # check the arguments
+  if (isTRUE(any(p <= 0 | p >= 1))) {
+    stop("quantile value has to be between 0 and 1")
+  }
+  if (isTRUE(any(sigma <= 0))) {
+    stop("generalized_normal is only defined for sigma > 0")
+  }
+  if (isTRUE(any(beta <= 0))) {
+    stop("generalized_normal is only defined for beta > 0")
+  }
+  if (!lenEqual(list_of_vectors = list(p, mu, sigma, beta), scalars_allowed = TRUE, type_check = is.numeric)) {
+    stop("generalized_normal argument vectors could not be matched. May be due to wrong type,
+         or different lengths. Note: len=1 is always allowed, even if the other vectors are len!=1.")
+  }
+
+  q_part <- ((sigma^beta) * qgamma(2*abs(p - 0.5), 1/beta))^(1/beta)
+  return(sign(p - 0.5) * q_part + mu)
+}
+
+#' RNG for the generalized_normal distribution
+#'
+#' @param n Number of sample
+#' @param mu Mean, unbound
+#' @param sigma Scale, sigma > 0
+#' @param beta shape, beta > 0
+#'
+#' @return Random numbers from the generalized_normal distribution.
+#' @export
+#'
+#' @examples hist(rgeneralized_normal(100, mu = 2, sigma = 2, beta = 2))
+rgeneralized_normal <- function(n, mu = 0, sigma = 1, beta = 1) {
+  # check the arguments
+  if (!isNat_len(n)) {
+    stop("The number RNG-samples has to be a scalar natural")
+  }
+  return(qgeneralized_normal(p = runif(n, min = 0, max = 1), mu = mu, sigma = sigma, beta = beta))
+}
+
+#' Log-Likelihood of the generalized_normal distribution
+#'
+#' @param i BRMS indices
+#' @param prep BRMS data
+#'
+#' @return Log-Likelihood of generalized_normal given data in prep
+log_lik_generalized_normal <- function(i, prep) {
+  mu <- brms::get_dpar(prep, "mu", i = i)
+  sigma <- brms::get_dpar(prep, "sigma", i = i)
+  beta <- brms::get_dpar(prep, "beta", i = i)
+  y <- prep$data$Y[i]
+  return(dgeneralized_normal(x = y, mu = mu, sigma = sigma, beta = beta, log = TRUE))
```

Comment:
We could at least test if all these run without error. I think that storing reference results for each would be a little too much.